README.md
# senseBox MCU Arduino Core Board Support Package [![Build Status](https://travis-ci.org/sensebox/arduino-senseBoxCore.svg?branch=master)](https://travis-ci.org/sensebox/arduino-senseBoxCore)

This repository contains the Board Support Package for the senseBox MCU required to compile Arduino sketches in [Arduino IDE v1.8+](https://www.arduino.cc/en/Main/Software).

To install the core in your Arduino IDE:

  1. Open the **Preferences** of the Arduino IDE.
  1. Add this URL `https://sensebox.github.io/arduino-senseBoxCore/package_sensebox_index.json` in the **Additional Boards Manager URLs** field, and click OK.
  1. Open the **Boards Manager** (menu Tools->Board->Board Manager...)
  1. Install **senseBox MCU Board Core**
  1. Select the **senseBox MCU** board under **senseBox** in Tools->Board menu
  1. Connect your senseBox MCU through USB
  1. Compile/Upload as usual

## Features

* [Microchip ATSAMD21](http://www.microchip.com/wwwproducts/en/ATSAMD21G18) ARM Cortex-M0+ Microcontroller
* [Microchip ATECC608A](http://www.microchip.com/wwwproducts/en/ATECC608A) Crypto Authentication
* [Bosch BMX055](https://www.bosch-sensortec.com/bst/products/all_products/bmx055) Orientation Sensor (Accelerometer + Gyroscope + Magnetometer)
* USB CDC+MSC Bootloader (Arduino compatible)
* XBee compatible sockets with UART and SPI
* 5V tolerant IOs, UART, I2C

## Hardware and Software

* [Schematics + Layout](https://github.com/watterott/SenseBox-MCU/tree/master/hardware) (External link)
* [Board Support Package for Arduino IDE](https://github.com/sensebox/arduino-senseBoxCore/)

## Try changes locally

This only works on linx!

1. `./extras/pack.release.bash`
1. `./extras/test.release.bash prepareJson`
1. If you have the current version already installed, delete the folder `~/.arduino15/packages/senseBox`
1. Run `arduino --pref "boardsmanager.additional.urls=file:///home/your-user/arduino-senseBoxCore/package_sensebox_index.json" --install-boards "senseBox:samd:${VERSION}"`

## Releasing a new version

Make sure there isn't already a release with this version on Github! The release provider will not overwrite the release but force push the json resulting in different crc values of the actual archive file and in the json!

1. Commit all of your changes
1. Increase the version in `platform.txt`
1. Try to build locally `./extras/pack.release.bash`
1. Run `./extras/pack.release.bash commitAndTag`
1. Run `git push origin <NEW_VERSION> master`
1. Travis CI will take over to:
    * Create the archive
    * Try to compile a basic sketch
    * Create a release containing the archive
    * Update the `package_sensebox_index.json` on the `gh-pages` branch

## Sources

[https://downloads.arduino.cc/packages/package_index.json](https://downloads.arduino.cc/packages/package_index.json)
[https://github.com/watterott/SAM-BAR](https://github.com/watterott/SAM-BAR/tree/67850505de43aa24ca12c3899b9348745af7809d)
[https://github.com/watterott/SenseBox-MCU](https://github.com/watterott/SenseBox-MCU)
[https://github.com/arduino/ArduinoCore-samd](https://github.com/arduino/ArduinoCore-samd)

## Changes made to the original Arduino Core for SAMD21 CPU

* Extended the `extras/pack.release.bash` file to create and deploy JSON and package to Github pages
* Replaced files in `bootloaders`, `drivers`and `variants` from with senseBox Board stuff from watterott/SenseBox-MCU
* Added senseBoxIO library to `libraries`
* Replaced `boards.txt` with senseBox Board definitions from watterott/SenseBox-MCU
* Changed `version` and `name` fields in `platform.txt`

## License and credits

This core is based on work by Arduino LL, Atmel and Watterott.

License for parts by Arduino LLC

    Copyright (c) 2015 Arduino LLC.  All right reserved.

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
    See the GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


CHANGELOG
1.0.2
 - Removed custom WiFi101 library

1.0.1
  - First Release for senseBox MCU

1.0.0 (not available)
 - Release for senseBox MCU prototype

extras/test.release.bash
#!/bin/bash
set -euo pipefail
IFS=$'\n\t'

VERSION=$(grep version= platform.txt | sed 's/version=//g')

BUILD_DIR=${TRAVIS_BUILD_DIR:-$(pwd)}

FILENAME="arduino-senseBoxCore-$VERSION.tar.bz2"
OUTPUTFOLDER="packages"
OUTPUTNAME="$OUTPUTFOLDER/$FILENAME"

JSON_FILENAME="package_sensebox_index.json"
JSON_FILENAME_BCKP="package_sensebox_index.json_original"
SENSEBOXCORE_URL="file:///${BUILD_DIR}/${JSON_FILENAME}"
BOARD_TO_INSTALL="senseBox:samd:${VERSION}"

BOARD="senseBox:samd:sb"

prepareJson () {
  echo "Creating testing package_sensebox_index.json ... "
  mv "${JSON_FILENAME}" "${JSON_FILENAME_BCKP}"
  # modify the package_sensebox_index.json to use local url for the package
  jq -rM --arg version "$VERSION" --arg url "file://${BUILD_DIR}/${OUTPUTNAME}" -f extras/replace_local_url.jq "${JSON_FILENAME_BCKP}" > "${JSON_FILENAME}"
  echo "done"
}

testRelease () {
  echo "Installing the core ... "
  arduino --pref "boardsmanager.additional.urls=${SENSEBOXCORE_URL}" --install-boards "${BOARD_TO_INSTALL}"
  echo "done"

  # Build the Blink example
  echo "Compiling test sketch ... "
  arduino --verbose-build --verify --board "${BOARD}" "${BUILD_DIR}/libraries/senseBox/examples/Blink/Blink.ino"
  echo "done"
}

cmd=${1:-}

if [[ -z "$cmd" ]]; then
  prepareJson
  testRelease
else
  eval "${cmd}"
fi


libraries/senseBoxIO/senseBoxIO.h
#define I2C_PWR 19  // PA05
#define SD_PWR 21   // PA07
#define UART_PWR 20 // PA06
#define XBEE1_PWR 5 // PB11
#define XBEE2_PWR 4 // PB10

#define MCU_CS_SD 28    // PA14
#define MCU_CS_XBEE1 24 // PA18
#define MCU_CS_XBEE2 6  // PA20

#include <Arduino.h>

class SenseBoxIO {
public:
  SenseBoxIO() {
    SPISelectNone();
    PowerI2C(false);
    PowerSDCard(false);
    PowerUART(false);
    PowerXBEE1(false);
    PowerXBEE2(false);
  }

  void PowerI2C(bool on) {
    pinMode(I2C_PWR, OUTPUT);
    digitalWrite(I2C_PWR,
                 (on) ? HIGH
                      : LOW); // turn Power on or off. Port is non inverted!
  }

  void PowerSDCard(bool on) {
    pinMode(SD_PWR, OUTPUT);
    digitalWrite(SD_PWR, (on) ? LOW : HIGH); // turn Power on or off
  }

  void PowerUART(bool on) {
    pinMode(UART_PWR, OUTPUT);
    digitalWrite(UART_PWR,
                 (on) ? HIGH
                      : LOW); // turn Power on or off. Port is non inverted!
  }

  void PowerXBEE1(bool on) {
    pinMode(XBEE1_PWR, OUTPUT);
    digitalWrite(XBEE1_PWR, (on) ? LOW : HIGH); // turn Power on or off
  }

  void PowerXBEE2(bool on) {
    pinMode(XBEE2_PWR, OUTPUT);
    digitalWrite(XBEE2_PWR, (on) ? LOW : HIGH); // turn Power on or off
  }

  void SPISelectNone(void) {
    pinMode(MCU_CS_XBEE1, OUTPUT);
    digitalWrite(MCU_CS_XBEE1, HIGH); // turn off

    pinMode(MCU_CS_XBEE2, OUTPUT);
    digitalWrite(MCU_CS_XBEE2, HIGH); // turn off

    pinMode(MCU_CS_SD, OUTPUT);
    digitalWrite(MCU_CS_SD, HIGH); // turn off
  }

  void SPISelectSDCard(void) {
    pinMode(MCU_CS_XBEE1, OUTPUT);
    digitalWrite(MCU_CS_XBEE1, HIGH); // turn off

    pinMode(MCU_CS_XBEE2, OUTPUT);
    digitalWrite(MCU_CS_XBEE2, HIGH); // turn off

    pinMode(MCU_CS_SD, OUTPUT);
    digitalWrite(MCU_CS_SD, LOW); // turn on
  }

  void SPISelectXBEE1(void) {
    pinMode(MCU_CS_XBEE2, OUTPUT);
    digitalWrite(MCU_CS_XBEE2, HIGH); // turn off

    pinMode(MCU_CS_SD, OUTPUT);
    digitalWrite(MCU_CS_SD, HIGH); // turn off

    pinMode(MCU_CS_XBEE1, OUTPUT);
    digitalWrite(MCU_CS_XBEE1, LOW); // turn on
  }

  void SPISelectXBEE2(void) {
    pinMode(MCU_CS_XBEE1, OUTPUT);
    digitalWrite(MCU_CS_XBEE1, HIGH); // turn off

    pinMode(MCU_CS_SD, OUTPUT);
    digitalWrite(MCU_CS_SD, HIGH); // turn off

    pinMode(MCU_CS_XBEE2, OUTPUT);
    digitalWrite(MCU_CS_XBEE2, LOW); // turn on
  }
};

static SenseBoxIO senseBoxIO;


libraries/senseBoxIO/src/SenseBoxIO.h
#include <Arduino.h>

#define PIN_I2C_PWR (5)
#define PIN_UART_PWR (16)
#define PIN_XB1_PWR (17)
#define PIN_XB2_PWR (4)
#define PIN_XB1_CS (24)
#define PIN_XB2_CS (28)
#define PIN_XB1_INT (7)
#define PIN_XB2_INT (29)
#define PIN_RED_LED (30)
#define PIN_GREEN_LED (31)
#define PIN_SWITCH (6)
#define PIN_IO1 (18)
#define PIN_IO2 (19)
#define PIN_IO3 (20)
#define PIN_IO4 (21)
#define PIN_IO5 (25)
#define PIN_IO6 (15)

#define I2C_ATECC (0x60)  // ATECC608A
#define I2C_ACCEL (0x18)  // BMX055
#define I2C_GYRO (0x68)   // BMX055
#define I2C_MAGNET (0x12) // BMX055

class SenseBoxIO {
public:
  SenseBoxIO() {
    pinMode(PIN_XB1_CS, OUTPUT);
    pinMode(PIN_XB2_CS, OUTPUT);
    pinMode(PIN_RED_LED, OUTPUT);
    pinMode(PIN_GREEN_LED, OUTPUT);
    PowerI2C(false);
    PowerUART(false);
    PowerXB1(false);
    PowerXB2(false);
    SPISelectNone();
    StatusNone();
  }

  void StatusNone(void) {
    digitalWrite(PIN_GREEN_LED, LOW);
    digitalWrite(PIN_RED_LED, LOW);
  }

  void StatusRed(void) {
    digitalWrite(PIN_GREEN_LED, LOW);
    digitalWrite(PIN_RED_LED, HIGH);
  }

  void StatusGreen(void) {
    digitalWrite(PIN_RED_LED, LOW);
    digitalWrite(PIN_GREEN_LED, HIGH);
  }

  void StatusSet(int s) {
    switch (s) {
    case 0:
      StatusNone();
      break;
    case 1:
      StatusRed();
      break;
    case 2:
      StatusGreen();
      break;
    case 3:
      StatusRed();
      StatusGreen();
      break;
    }
  }

  void PowerI2C(bool on) {
    pinMode(PIN_I2C_PWR, OUTPUT);
    digitalWrite(PIN_I2C_PWR, (on) ? HIGH : LOW); // non inverted
  }

  void PowerUART(bool on) {
    pinMode(PIN_UART_PWR, OUTPUT);
    digitalWrite(PIN_UART_PWR, (on) ? HIGH : LOW); // non inverted
  }

  void PowerXB1(bool on) {
    pinMode(PIN_XB1_PWR, OUTPUT);
    digitalWrite(PIN_XB1_PWR, (on) ? LOW : HIGH); // inverted
    if (on) {
      delay(500); // wait 500ms (MAX809 reset circuit)
    }
  }

  void PowerXB2(bool on) {
    pinMode(PIN_XB2_PWR, OUTPUT);
    digitalWrite(PIN_XB2_PWR, (on) ? LOW : HIGH); // inverted
    if (on) {
      delay(500); // wait 500ms (MAX809 reset circuit)
    }
  }

  void SPISelectNone(void) {
    digitalWrite(PIN_XB1_CS, HIGH);
    digitalWrite(PIN_XB2_CS, HIGH);
  }

  void SPISelectXB1(void) {
    digitalWrite(PIN_XB2_CS, HIGH);
    digitalWrite(PIN_XB1_CS, LOW);
  }

  void SPISelectXB2(void) {
    digitalWrite(PIN_XB1_CS, HIGH);
    digitalWrite(PIN_XB2_CS, LOW);
  }
};

static SenseBoxIO senseBoxIO;


extras/pack.release.bash
#!/bin/bash

#  pack.*.bash - Bash script to help packaging samd core releases.
#  Copyright (c) 2015 Arduino LLC.  All right reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 2.1 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

set -euo pipefail
IFS=$'\n\t'

# check if needed tools are installed
jq --version >/dev/null 2>&1 || { echo >&2 "This script requires jq but it's not installed.  Aborting."; exit 1; }
curl --version >/dev/null 2>&1 || { echo >&2 "This script requires curl but it's not installed.  Aborting."; exit 1; }
tar --version >/dev/null 2>&1 || { echo >&2 "This script requires tar but it's not installed.  Aborting."; exit 1; }

VERSION=$(grep version= platform.txt | sed 's/version=//g')

PWD=$(pwd)
FOLDERNAME=$(basename "$PWD")
FILENAME="arduino-senseBoxCore-$VERSION.tar.bz2"
OUTPUTFOLDER="packages"
OUTPUTNAME="$OUTPUTFOLDER/$FILENAME"
INPUTNAME=$(dirname "$PWD")

# These variables hold references to the original Arduino boards package we are
# based on
ARDUINOBOARDPACKAGEURL="https://downloads.arduino.cc/packages/package_index.json"
ARDUINOBOARDPACKAGENAME="Arduino SAMD Boards (32-bits ARM Cortex-M0+)"
ARDUINOBOARDPACKAGEVERSION="1.6.17"

# wrap in functions to ensure script can be loaded correctly before executing anything
createArchive () {
  mkdir -p "$OUTPUTFOLDER"
  if [ -f "$OUTPUTNAME" ]; then
    BACKUPNAME="${OUTPUTNAME}_bckp"
    echo "File ${OUTPUTNAME} already exists. Renaming to ${BACKUPNAME}"
    mv "${OUTPUTNAME}" "${BACKUPNAME}"
  fi

  echo -n "Creating tar archive $OUTPUTNAME ... "

  tar -C "$INPUTNAME" --transform "s|$FOLDERNAME|$FOLDERNAME-$VERSION|g" --exclude=extras --exclude=.git* --exclude=packages --exclude package_sensebox_index.json --exclude-from=.gitignore -cjf "$OUTPUTNAME" "$FOLDERNAME"

  echo "done"
}

updatePackageJson () {
  echo "Updating package_sensebox_index.json ... "

  echo -en "\\tGetting toolsDependencies from upstream package ... "

  TOOLSDEPENDENCIES=$(curl --silent -L "$ARDUINOBOARDPACKAGEURL" | jq -rMc --arg name "$ARDUINOBOARDPACKAGENAME" --arg version "$ARDUINOBOARDPACKAGEVERSION" -f extras/extract_toolsDependencies.jq)

  echo "done"

  echo -en "\\tGetting previous package_sensebox_index.json from gh-pages branch ... "

  curl -o package_sensebox_index.json --silent -L "https://sensebox.github.io/arduino-senseBoxCore/package_sensebox_index.json"

  echo "done"

  TRAVIS_IS_BUILDING_TAG=${TRAVIS_TAG:-}

  if [[ -z "${TRAVIS_IS_BUILDING_TAG}" ]]; then
    echo -en "\\t\\tRemoving version ${VERSION} from package_sensebox_index.json since this is not a new tag ... "

    jq -rM --arg version "${VERSION}" -f extras/remove_currentPlatform.jq package_sensebox_index.json > new_package.json
    mv new_package.json package_sensebox_index.json

    echo "done"
  fi

  echo -en "\\tConstructing new platform ... "

  # get filesize
  FILESIZE=$(stat -c "%s" "$OUTPUTNAME")

  # calculate checksum
  CHECKSUM="SHA-256:$(shasum -a 256 "$OUTPUTNAME" | cut -d " " -f 1)"

  NEWPLATFORM=$(jq -rMc --argjson toolsDeps "$TOOLSDEPENDENCIES" --arg version "$VERSION" --arg checksum "$CHECKSUM" --arg filesize "$FILESIZE" --arg filename "$FILENAME" -f extras/newPlatform.jq extras/platform.json.template)

  echo "done"

  echo -en "\\tConstructing updated package_sensebox_index.json ... "

  jq -rM --argjson newPlatform "$NEWPLATFORM" -f extras/update_package.jq package_sensebox_index.json > new_package.json
  mv new_package.json package_sensebox_index.json

  echo "done"

  echo "done"
}

commitAndTag () {
  echo -n "Creating commit ... "
  git commit -m "Release ${VERSION}"
  echo "done"

  echo -n "Creating new tag ${VERSION} ... "
  git tag "${VERSION}"
  echo "done"

  echo ""
  echo "Now run"
  echo -e "\\tgit push origin ${VERSION} master"
}

cmd=${1:-}

if [[ -z "$cmd" ]]; then
  createArchive
  updatePackageJson
else
  eval "${cmd}"
fi
